src/Kestrel.Core/Internal/HttpsConnectionAdapter.cs

```diff
@@ -253,5 +274,84 @@ public void Dispose()
+
+            public PrefixedStream(Stream innerStream, Memory<byte> prefix)
+            {
+                _innerStream = innerStream;
+                _prefix = prefix;
+            }
+
+            public override bool CanRead => _innerStream.CanRead;
+            public override bool CanSeek => _innerStream.CanSeek;
+            public override bool CanWrite => _innerStream.CanWrite;
+            public override long Length => _innerStream.Length;
+            public override long Position { get => _innerStream.Position; set => _innerStream.Position = value; }
+
+            public override void SetLength(long value) => _innerStream.SetLength(value);
+            public override long Seek(long offset, SeekOrigin origin) => _innerStream.Seek(offset, origin);
+
+            public override void Flush() => _innerStream.Flush();
+            public override Task FlushAsync(CancellationToken cancellationToken) => _innerStream.FlushAsync(cancellationToken);
+
+            public override void Write(byte[] buffer, int offset, int count) => _innerStream.Write(buffer, offset, count);
+
+            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
+                _innerStream.WriteAsync(buffer, offset, count, cancellationToken);
+
+            public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
+                => _innerStream.BeginWrite(buffer, offset, count, callback, state);
+
+            public override void EndWrite(IAsyncResult asyncResult) => _innerStream.EndWrite(asyncResult);
+
+            protected override void Dispose(bool disposing) => _innerStream.Dispose();
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                if (!_prefix.IsEmpty)
+                {
+                    // Validates args
+                    var destination = new Span<byte>(buffer, offset, count);
+                    return ReadPrefix(destination);
+                }
+
+                return _innerStream.Read(buffer, offset, count);
+            }
+
+            public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                if (!_prefix.IsEmpty)
+                {
+                    // Validates args
+                    var destination = new Span<byte>(buffer, offset, count);
+                    return Task.FromResult(ReadPrefix(destination));
+                }
+
+                return _innerStream.ReadAsync(buffer, offset, count, cancellationToken);
+            }
+
+            private int ReadPrefix(Span<byte> destination)
+            {
+                if (destination.Length == 0)
+                {
+                    // Data is available
+                    return 0;
+                }
+
+                var read = Math.Min(destination.Length, _prefix.Length);
```

Comment:
Since `destination.Length != 0` and `_prefix.Length == 1` this isn't necessary.

Reply:
Nvm. future proofing I see. I agree we probably wan't to try to get a full TCP frame on the first read if we do this.